source/lac/petsc_parallel_block_sparse_matrix.cc

```diff
@@ -123,24 +155,63 @@ namespace PETScWrappers
     void
     BlockSparseMatrix::collect_sizes()
     {
-      BaseClass::collect_sizes();
+      auto           m = this->n_block_rows();
+      auto           n = this->n_block_cols();
+      PetscErrorCode ierr;
+
+      // Create empty matrices if needed
+      // This is neeeded by the base class
+      // not by MATNEST
+      std::vector<size_type> row_sizes(m);
+      std::vector<size_type> col_sizes(n);
+      std::vector<size_type> row_local_sizes(m);
+      std::vector<size_type> col_local_sizes(n);
+      MPI_Comm               comm = MPI_COMM_NULL;
+      for (size_type r = 0; r < m; r++)
+        {
+          for (size_type c = 0; c < n; c++)
+            {
+              if (this->sub_objects[r][c])
+                {
+                  comm = this->sub_objects[r][c]->get_mpi_communicator();
+                  row_sizes[r]       = this->sub_objects[r][c]->m();
+                  col_sizes[c]       = this->sub_objects[r][c]->n();
+                  row_local_sizes[r] = this->sub_objects[r][c]->local_size();
+                  col_local_sizes[c] =
+                    this->sub_objects[r][c]->local_domain_size();
+                }
+            }
+        }
```

Comment:
well, this is not actually a copy of the base code. It checks if any block is NULL and replaces it with a dummy matrix. Deal.II block matrix class does not support empty blocks as `nullptr`. `MATNEST` instead supports it. The alternative would be to modify the base block matrix class in many places.  I have chosen to implement these few extra lines here.

Reply:
I see. 

Can I ask you for one more change in this function then? You skip over blocks that are empty, but in order to determine the sizes correctly, you will have to have one non-empty block per block row and per block columns. This would be useful to assert at the end. To do this, I would initialize `row_sizes` and `row_columns` with `size_type(-1)`, and then after the double loop here, you'd make sure that the two arrays don't have any of these values left.

While you're there: You're repeating assignments to `row_sizes[r]` and `col_sizes[c]`, but in repeated assignments it would be useful to check that the current block's size matches what we had previously recorded via an assert. It is easy to get block sizes wrong when creating block matrices with rectangular blocks.

Reply:
All the checks you are asking for are already implemented in the base class `collect_sizes()`. I have opted for the easiest implementation here since any mistake will be spotted later.

Reply:
OK, but not quite. Below you have this code:
```
     for (size_type r = 0; r < m; r++)
￼        {
￼          for (size_type c = 0; c < n; c++)
￼            {
￼              if (!this->sub_objects[r][c])
￼                {
￼                  Mat dummy = ::create_dummy_mat(
￼                    comm,
￼                    static_cast<PetscInt>(row_local_sizes[r]),
￼                    static_cast<PetscInt>(row_sizes[r]),
￼                    static_cast<PetscInt>(col_local_sizes[c]),
￼                    static_cast<PetscInt>(col_sizes[c]));
```
This is going to fail if someone only provided empty blocks for a whole row or column -- in that case, we simply don't know what `row_sizes[r]` are, for example.

Reply:
@bangerth I see what you mean. I have added the size_type(-1) initialization. Thanks

Reply:
Right, but then below where you call `create_dummy_mat()`, you still have to assert that the sizes are valid.

Reply:
PETSc will complain, raise an error and the Assert inside create_dummy_mat will trigger the exception.